StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1210UsingDirectivesMustBeOrderedAlphabeticallyByNamespace.cs

```diff
@@ -50,108 +56,60 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
-                {
-                    continue;
-                }
+        private void HandleNamespaceDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            var namespaceDeclaration = (NamespaceDeclarationSyntax)context.Node;
 
-                string precedingFullyQualifiedNamespace = precedingNamespaceSymbol.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
-                if (string.Compare(fullyQualifiedNamespace, precedingFullyQualifiedNamespace, StringComparison.OrdinalIgnoreCase) >= 0)
-                {
-                    continue;
-                }
+            ProcessUsings(namespaceDeclaration.Usings, context);
+        }
 
-                string @namespace = namespaceSymbol.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat);
-                string precedingNamespace = precedingNamespaceSymbol.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat);
+        private static void ProcessUsings(SyntaxList<UsingDirectiveSyntax> usings, SyntaxNodeAnalysisContext context)
+        {
+            var usingsDirectivesWithoutAliasAndStatic = usings.Where(ExcludeAliasAndStaticUsingDirectives);
+            var systemUsingDirectives = usingsDirectivesWithoutAliasAndStatic.Where(directive => directive.IsSystemUsingDirective());
+            var usingDirectives = usingsDirectivesWithoutAliasAndStatic.Where(directive => !directive.IsSystemUsingDirective());
 
-                // Using directive for '{namespace}' must appear before directive for '{precedingNamespace}'
-                context.ReportDiagnostic(Diagnostic.Create(Descriptor, syntax.GetLocation(), @namespace, precedingNamespace));
-                break;
-            }
+            CheckOrderAndReportDiagnostic(usingDirectives.ToList(), context);
+            CheckOrderAndReportDiagnostic(systemUsingDirectives.ToList(), context);
         }
 
-        private static string GetTopLevelNamespace(SemanticModel semanticModel, UsingDirectiveSyntax syntax, out INamespaceSymbol namespaceSymbol, CancellationToken cancellationToken)
+        private static void CheckOrderAndReportDiagnostic(IList<UsingDirectiveSyntax> usings, SyntaxNodeAnalysisContext context)
         {
-            SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(syntax.Name, cancellationToken);
-            namespaceSymbol = symbolInfo.Symbol as INamespaceSymbol;
-            if (namespaceSymbol == null)
+            if (usings.Count <= 1)
             {
-                return null;
+                return;
             }
 
-            string fullyQualifiedName = namespaceSymbol.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
-            string name = fullyQualifiedName;
+            var properPositionsOfUsingDirectives = usings.OrderBy(directive => directive.Name.ToUnaliasedString()).Select((directive, idx) => new { Directive = directive, Index = idx }).ToDictionary(item => item.Directive, item => item.Index);
```

Comment:
Yeah, I missed that. I think we should stop analyzing at the first occurrence of SA1210.
I suggest following implementation.
```c#
private static void CheckOrderAndReportDiagnostic(SyntaxNodeAnalysisContext context, List<UsingDirectiveSyntax> usings)
{
    if (usings.Count <= 1)
    {
        return;
    }

    UsingDirectiveSyntax prevUsingDirective = null;
    var usingsNames = new List<string>(usings.Count);

    foreach (var usingDirective in usings)
    {
        string currentDirectiveName = usingDirective.Name.ToUnaliasedString();
        if (prevUsingDirective != null)
        {
            string prevDirectiveName = prevUsingDirective.Name.ToUnaliasedString();
            if (string.CompareOrdinal(prevDirectiveName, currentDirectiveName) > 0)
            {
                // Find directive before which should be placed current directive
                foreach (string usingName in usingsNames)
                {
                    if (string.CompareOrdinal(usingName, currentDirectiveName) > 0)
                    {
                        prevDirectiveName = usingName;
                        break;
                    }
                }

                // Using directive for '{currentDirectiveName}' must appear before directive for '{prevDirectiveName}'
                context.ReportDiagnostic(Diagnostic.Create(Descriptor, usingDirective.GetLocation(), currentDirectiveName, prevDirectiveName));
                return;
            }
        }

        usingsNames.Add(currentDirectiveName);
        prevUsingDirective = usingDirective;
    }
}
```
I intentionally avoid LINQ because it requires closure allocations.